### ISSUE TITLE
convenience api for demuxing all frames in one go

### ISSUE THREAD

**Issue by zenkyuv** [ForeverSc/web-demuxer#1]
could there be convenience api for demuxing all frames in one go ?
i see there is api for getting all packets (getAVPackets, if im correct), would be cool if there was method exposed to get all encoded frames and push them to decoder queue as they are demuxed (ideally both in their workers, if its possible to have demuxer in web worker?), not a big deal though, i havent tested your library yet, i will check it in my project soon (omniclip) and replace mp4boxjs, thanks for working on the library 👍

**Comment by ForeverSc** [comment]
you could use readAVPacket to get all frames, and demuxer is working in web worker. you can see the [readAVPacket example code](https://github.com/ForeverSc/web-demuxer/blob/main/index.html#L124-L164), which used to play video. 
because get all frames in one go maybe cost too much memory when file is large, so i only supported readAVPacket to get packets as a ReadableStream. But if you want get all frames api at once, i could add one.
Btw, omniclip is pretty cool ! 😁

**Comment by zenkyuv** [comment]
yea im aware its kind of bad on memory to get all frames at once, seems like the bottleneck is that file is being loaded whole at once, i could seek frame by frame if it is not causing memory spikes (if its just loading that only needed chunk of file to get that frame), although it will be probably slower than just demuxing whole file chunk by chunk without seeking to specific frame, because i dont need to seek, its for rendering process of my video editor where it just goes all one by one, anyways, do you think its possible to demux whole file but without loading whole file at once but kind of spread the loading of file in time? or load some small chunk of that file, and after that chunk is done demuxing then load another chunk of that file and demux and so on? Something to offset that big initial spike?

**Comment by ForeverSc** [comment]
I thought video edit and video export should be different. When editing video, i don't think you need get all frames at once, because you only need to edit the current frame. you can use seek frame to demux and edit the current frame. When exporting video, you can use readAVPacket to demux all frames one by one,  and replace the edited frame，then encode frame into new video file. only process one frame at one time,  will not cause so much memory.
readAVPacket default is to read avpacket one by one, or you can specify the start and end time to read.

**Comment by zenkyuv** [comment]
> I thought video edit and video export should be different. When editing video, i don't think you need get all frames at once, because you only need to edit the current frame. you can use seek frame to demux and edit the current frame. When exporting video, you can use readAVPacket to demux all frames one by one, and replace the edited frame，then encode frame into new video file. only process one frame at one time, will not cost so much memory. readAVPacket default is to read avpacket one by one, or you can specify the start and end time to read.

yep, i meant video export (i call it render process sometimes), could you give an example on how to demux file, without seeking, just getting all the frames, one by one ? seems like readAVPacket returns just one WebAVPacket, maybe you meant to use getAVPackets ?

**Comment by ForeverSc** [comment]
use readAVPacket to get all packets one by one:
```js
async function getAllPackets(file) {
  await demuxer.load(file)
  return new Promise((resolve) => {
    const videoPackets = []
    const reader = demuxer.readAVPacket().getReader()

    reader.read().then(async function processAVPacket({ done, value }) {
      if (done) {
        resolve(videoPackets) // all packets read finished
        return
      }

      videoPackets.push(value) // get one video packet
      // and you can generate video chunk for video decoder
      // const videoChunk = demuxer.genEncodedVideoChunk(value)

      return reader.read().then(processAVPacket)
    })
  })
}
```
updated the [example](https://foreversc.github.io/web-demuxer/#example-get-all-video-packets)

**Comment by zenkyuv** [comment]
> use readAVPacket to get all packets one by one:
> 
> ```js
> async function getAllPackets(file) {
>   await demuxer.load(file)
>   return new Promise((resolve) => {
>     const videoPackets = []
>     const reader = demuxer.readAVPacket().getReader()
> 
>     reader.read().then(async function processAVPacket({ done, value }) {
>       if (done) {
>         resolve(videoPackets) // all packets read finished
>         return
>       }
> 
>       videoPackets.push(value) // get one video packet
>       // and you can generate video chunk for video decoder
>       // const videoChunk = demuxer.genEncodedVideoChunk(value)
> 
>       return reader.read().then(processAVPacket)
>     })
>   })
> }
> ```
> 
> updated the [example](https://foreversc.github.io/web-demuxer/#example-get-all-video-packets)

thanks, it works ! I think the issue can be closed then, seems like the convenience api is not needed at all
